### PR TITLE
Fix ldiv / rdiv + implement pinv

### PR DESCRIPTION
**Pull request by willtebbutt**
Relates to #222

### CODE REVIEW

File: src/lib/array.jl [hunk 1]
```diff
@@ -204,18 +214,55 @@ end
...
   Y = A \ B
   return Y, function(Ȳ)
-      B̄ = A' \ Ȳ
-      return (-B̄ * Y', B̄)
+    B̄ = A' \ Ȳ
+    return (-B̄ * Y', B̄)
+  end 
+end
+
+@adjoint function /(A::AbstractMatrix, B::Union{Diagonal, AbstractTriangular})
```
Comment (ararslan) [inline src/lib/array.jl]: What about `/` for abstract matrices? Is that still covered and I'm missing it?

File: src/lib/array.jl [hunk 2]
```diff
@@ -204,18 +214,55 @@ end
...
+    Ā = Ȳ / B'
+    return (Ā, -Y' * Ā)
+  end
+end
+
+@adjoint function \(A::AbstractMatrix, B::AbstractVecOrMat)
+  Z = A \ B
+  return Z, function(Z̄)
+    B̄ = A' \ Z̄
+    if size(A, 1) == size(A, 2)
```
Comment (ararslan) [inline src/lib/array.jl]: Don't we also need that `A` is nonsingular?

File: src/lib/array.jl (hunk 1 above)
Comment (willtebbutt) [reply src/lib/array.jl]: Yup. `/` is implemented in terms of `\` for abstract matrices :)

File: src/lib/array.jl (hunk 2 above)
Comment (willtebbutt) [reply src/lib/array.jl]: I just elected to follow the structure in base found [here](https://github.com/JuliaLang/julia/blob/5d02c59185ef31703891adc748e0ed414a9a5b3f/stdlib/LinearAlgebra/src/generic.jl#L956) precisely. I think base assumes that the user will ensure that `A` is well-conditioned.

**Comment by MikeInnes** [comment]
LGTM. If you need the adjoint/transpose fixes can you rebase or pick the commits from #216, so we can include the credit?

**Comment by willtebbutt** [comment]
> If you need the adjoint/transpose fixes can you rebase or pick the commits from #216, so we can include the credit?

Well that's uncanny. I hadn't realised #216 is a thing, and somehow we've managed to implement pretty much the same thing character-for-character... Have cherry-picked both commits (not done this before, so may have made a mess of it :grimacing:)

**Comment by MikeInnes** [comment]
Funny, yeah they really do seem identical. Once tests pass I think this is good to go.

**Pull request merged**